### ISSUE TITLE
Multiple EXC_BAD_ACCESS Crashes

### ISSUE THREAD

**Issue by kunalsood** [realm/realm-cocoa#6730]
## Goals
Have my app run normally without crashes.

## Expected Results
Same as "Goals" (above).

## Actual Results
I'm seeing EXC_BAD_ACCESS in at least 2 different places. Please see:-
![](https://user-images.githubusercontent.com/1951423/91628316-1acde700-e9dc-11ea-90b8-2624132f8651.png)
![](https://user-images.githubusercontent.com/1951423/91628325-2caf8a00-e9dc-11ea-9a6a-c18c7857190f.png)

This might be related to [#6556](https://github.com/realm/realm-cocoa/issues/6556), but since that has been closed for a while, I decided to open a new issue for this.

## Steps for others to Reproduce
So far, I have been unable to reproduce this in a sample project, but, here are a few things I can tell:-

* Crash occurs randomly during a mass import when several write transactions are committed in quick succession on a single serial operation queue.
* This happens on brand new installation (i.e. not upgrading from old file format, also, not following compaction).
* My realm is stored in shared app group container (if that matters).
* So far, I have been unable to reproduce this on the Simulator, but it occurs fairly frequently when running on an actual device (iPad 6th generation).
* Since one of the crashes happens in `do_encryption_read_barrier`, I think it's also worth mentioning that my Realm is not encrypted.
* ~I have noticed that I can get rid of this crash if is slow down my writes (since my writes are committed on a serial background queue using NSOperation subclasses, adding `[NSThread sleepForTimeInterval:1.0];` after `[realm commitWriteTransaction];` in my NSOperation subclass is enough, but this is obviously not ideal).~ I eventually hit this crash irrespective on this change.
* One way that I have found to get rid of these crashes is to not use realm notifications of any kind, or use KVO on realm objects. (Which is also not ideal).
* A pattern I have noticed: This crash manifests itself after I have crashed once with a "mmap() failed: Cannot allocate memory size: XXXXXXXX offset: 0" exception during a mass import upon a fresh install, after which there are only EXC_BAD_ACCESS crashes no "mmap() failed" exceptions. I must say though: the exact same code had absolutely no "mmap() failed" exceptions or EXC_BAD_ACCESS crashes when used with realm-cocoa version 4.3.2.

## Code Sample

None right now.

## Version of Realm and Tooling

Realm framework version: 5.3.5

Realm Object Server version: N/A

Xcode version: 12 Beta 6

iOS/OSX version: iOS 14 Beta 6

Dependency manager + version: SPM (or Dynamic Framework built from source)

**Comment by leemaguire** [comment]
Hi @kunalsood ,

Thanks for the detail in your issue template! Would you be able to share what your Realm models look like and what size mass import your are doing?

Thanks,
Lee

**Comment by kunalsood** [comment]
> Would you be able to share what your Realm models look like

Sure thing (It's an ObjC Project, so sharing what my headers look like) :- 
```
@interface Feed : RLMObject

@property NSString *uniqueIdentifier; //Primary Key Property
@property NSString *title;
@property NSString *url;
@property NSString *displayTitle;
@property NSString *siteUrl;
@property NSString *summary;

@property NSInteger alphabeticalSortingDisplayIndex;

@property (readonly) RLMLinkingObjects *folders;
@property (readonly) RLMLinkingObjects *articles;

@end 

@interface Article : RLMObject

@property NSString *uniqueIdentifier; //Primary Key Property
@property NSString *syncServiceIdentifier;
@property NSString *title;
@property NSString *link;
@property NSString *image;
@property NSString *excerpt;
@property NSString *author;

@property (readonly) NSString *summary; 
@property (readonly) NSString *content;

@property double date;

@property BOOL isStarred;
@property BOOL isUnread;

@property Feed *feed;



@interface Folder : RLMObject

@property NSString *uniqueIdentifier; //Primary Key Property
@property NSString *displayName;

@property NSInteger alphabeticalSortingDisplayIndex;

@property RLMArray<Feed> *feeds;

@end



@interface Articles : RLMObject

//There should only ever be one object of this kind in the database

@property NSString *uniqueIdentifier; //Primary Key Property 

@property RLMArray<Article> *oldestFirst;
@property RLMArray<Article> *newestFirst;

@end
```

This should probably be self-explanatory in telling what properties are stored in realm & what aren't, and which ones are relationships & which ones are inverse-relationships. Please let me know if you'd also like to see my implementation (i.e. what is in my '.m' files) of these.

> what size mass import your are doing?

I'm importing no more than 15-20 `Folder` objects, 30-40 `Feed` objects, 6000-50000 `Article` objects, and only 1 `Articles` object. The order in which my mass import goes is as follows:- 
Step 1: Create a unique `Articles` object is it doesn't already exist.
Step 2: Create all `Feed` objects in a single write transaction.
Step 3: Create all `Folder` objects, and establish relationships with appropriate `Feed` objects in a single write transaction.
Step 4: Create `Article` objects in batches of 100, establishing relationships with appropriate `Feed` objects as I go though them. So, for 6000 `Article`objects, I would end of committing 60 write transactions.

With Realm 5.3.5, I'm not able to get to the end of this import without crashing (4 out of 5 times), even on my smallest test case of 6000 `Article` objects.

**Comment by finnschiermer** [comment]
This could be explained by a bug where we run notifications after the relevant realm version has been reclaimed and is no longer accessible.

**Comment by kunalsood** [comment]
> This could be explained by a bug where we run notifications after the relevant realm version has been reclaimed and is no longer accessible.

That would make sense assuming the way this is done in core has changed.

**Comment by kunalsood** [comment]
**UPDATE**: In my (completely anecdotal/unsubstantiated) testing with Realm Cocoa 5.4.0, I have been unable to re-produce this crash. I will post another update in the next few days.

**Comment by kunalsood** [comment]
**UPDATE**: Still hitting this crash on 5.4.0

**Comment by kunalsood** [comment]
**UPDATE**: Still hitting this crash on 5.4.7, also, hitting this crash even if I'm not using Realm's notifications, or KVO (this might have been the case al along, just today is the first time I've hit this crash when not using notifications or KVO).

**Comment by kunalsood** [comment]
It may also be worth noting that all these crashes happen on the main thread. I do not write to Realm from the main thread ever. So, this probably has to do with some kind of corruption happening in the database when writing to it from an NSOperationQueue, which causes these crashes when trying to read from the realm after said writes. (Of course, I'm just speculating this).

**Comment by finnschiermer** [comment]
@kunalsood Thx for info. 

1) Can you provide a stack backtrace of a crash for the case where you do not use KVO or notifications at all? (if it's different from the one at the top of this issue), it might help us.

2)  You wrote at the top of the issue:
~~~
A pattern I have noticed: This crash manifests itself after I have crashed once with a "mmap() failed: Cannot allocate memory size: XXXXXXXX offset: 0" exception during a mass import upon a fresh install, after which there are only EXC_BAD_ACCESS crashes no "mmap() failed" exceptions. I must say though: the exact same code had absolutely no "mmap() failed" exceptions or EXC_BAD_ACCESS crashes when used with realm-cocoa version 4.3.2.
~~~
Is this still correct?

Also, I need to be sure I interpret this correctly: you are saying that first the app runs fine (importing lots of data) until it crashes with mmap failed. Then subsequent runs of the app fails immediately (or fast) with EXC_BAD_ACCESS ?  At the point of the mmap failed, what's the size of the realm file? Is it possible to get a stack backtrace at this point (as opposed to at the EXC_BAD_ACCESS) ?

**Comment by kunalsood** [comment]
> @kunalsood Thx for info.
> 
> 1. Can you provide a stack backtrace of a crash for the case where you do not use KVO or notifications at all? (if it's different from the one at the top of this issue), it might help us.

It crashed at ```realm::util::EncryptedFileMapping::read_barrier```, I do not have a stack trace handy right now, but, I'll try and get one to you over the new few days.

> 2. You wrote at the top of the issue:
> 
> "A pattern I have noticed: This crash manifests itself after I have crashed once with a "mmap() failed: Cannot allocate memory size: XXXXXXXX offset: 0" exception during a mass import upon a fresh install, after which there are only EXC_BAD_ACCESS crashes no "mmap() failed" exceptions. I must say though: the exact same code had absolutely no "mmap() failed" exceptions or EXC_BAD_ACCESS crashes when used with realm-cocoa version 4.3.2."
> 
> Is this still correct?

Yes, this is still correct.

> 
> Also, I need to be sure I interpret this correctly: you are saying that first the app runs fine (importing lots of data) until it crashes with mmap failed. Then subsequent runs of the app fails immediately (or fast) with EXC_BAD_ACCESS ? At the point of the mmap failed, what's the size of the realm file? Is it possible to get a stack backtrace at this point (as opposed to at the EXC_BAD_ACCESS) ?
> 

The import never completes. At first attempted import, it crashes with mmap failed at some random point/line-of-my-code during the import (it is at a different point/line-of-my-code every time). When I re-launch the app, and attempt the import again, it eventually crashes again with EXC_BAD_ACCESS (again, at different point/line-of-my-code every time).
Does that clarify it? Please let me know if it doesn't.

> 1. Do you at any point have multiple processes (not just threads) accessing the realm file?

Yes, I do have iOS 14 Widgets reading (never writing) from the realm file at certain points (I call appropriate WidgetCenter APIs to request the system to reload the widget data, but, when the system actually attempts a reload is not something I can control, as the system throttles reload attempts to conserve battery and other resources etc.)

**Comment by finnschiermer** [comment]
Ok, last thing first:

"Yes, I do have iOS 14 Widgets reading (never writing) from the realm file at certain points (I call appropriate WidgetCenter APIs to request the system to reload the widget data, but, when the system actually attempts a reload is not something I can control, as the system throttles reload attempts to conserve battery and other resources etc.)"

Does such a widget have the realm file open... if so, how often does it refresh its view of the data? Or does it work by only opening the realm when it needs access, reading data, then closing it again immediately? What I'm after is: could a widget lock on to a version of realm data for an extended period of time, while the app is rapidly committing data as it receives it from the server?

**Comment by finnschiermer** [comment]
Also, checking my understanding: We have at least 3 active entities working in parallel here:
1) A widget. It only reads from the database.
2) The main thread. It also only reads from the database.
3) A background thread. It writes heavily.
And it is always the main thread that crashes.

**Comment by kunalsood** [comment]
> Does such a widget have the realm file open... if so, how often does it refresh its view of the data? Or does it work by only opening the realm when it needs access, reading data, then closing it again immediately? What I'm after is: could a widget lock on to a version of realm data for an extended period of time, while the app is rapidly committing data as it receives it from the server?

From my understand of how widgets work, no, widgets should not lock on to a version of the data in realm for an extended period of time at all. It only reads from the database when it needs to create an updated view, and does not hold on to any realm, realm objects, realm collection objects, or notification tokens. I just read from the realm, convert the data I need from realm to native objects, and let the realm go. This also does not happen very frequently during the import either. At most 2-3 times.

**Comment by kunalsood** [comment]
> Also, checking my understanding: We have at least 3 active entities working in parallel here:
> 
> 1. A widget. It only reads from the database.
> 2. The main thread. It also only reads from the database.
> 3. A background thread. It writes heavily.
> 
> And it is always the main thread that crashes. The widget never crashes. Are there other members at this party?

Yes, that is all correct. No other members at this party.

**Comment by RealmBot** [comment]
➤ Finn Andersen commented:

Thx a lot, just to rule out some possibilities: can you tell me the size of the realm file and (especially) the size of the lock file ?

**Comment by kunalsood** [comment]
> ➤ Finn Andersen commented:
> 
> Thx a lot, just to rule out some possibilities: can you tell me the size of the realm file and (especially) the size of the lock file ?

It'll take me some time to get you this info, as I do not use Realm 5.x in production (due to this exact issue), but, I'll get you this info APAP. Apologies about the delay.

**Comment by kunalsood** [comment]
Here we go:

```
file extension: realm
file byte size count: 335544320
formatted file size: 335.5 MB

file extension: lock
file byte size count: 1184
formatted file size: 1 KB


file extension: note
file byte size count: 0
formatted file size: Zero KB
```

This is using Realm Cocoa version 5.4.8
This particular crash was hit when about 45k into an import of around 55k `Article` objects.

Just as an aside: I do not appear to be hitting this crash as often as I used to. What I mean when I say that is: I was able to successfully complete my import a few times before I hit this crash. I only hit this crash when using KVO & Realm Notifications.

**Comment by jsflax** [comment]
@kunalsood can we confirm that this is still happening on the latest 10.x version of Realm?

**Comment by kunalsood** [comment]
I'm afraid I won't have a chance to test this throughly enough before January. So, please feel free to close this issue if you must, and I'll open a new issue if I see this when I get around to testing it.

**Comment by jsflax** [comment]
That's fair, we appreciate the heads up– we'll close this for now and do reopen it if it persists on the latest versions of Realm.

**Comment by ekimia** [comment]
update here: Ive noticed the same issue if I'm integrating realm via SPM. Using carthage is just fine.
